### ISSUE TITLE
Usage with blob strings / binary values

### ISSUE THREAD

**Issue by prat0088** [mgravell/Bedrock.Resp#1]
I've tried out Respite today and so far so good!  That example in the readme is a life saver.  One comment and one question:

1. I needed a way to write a block of memory as an array value, so I made this Create() a public method. I'm pretty sure that's not the best way, but it's working for now.
2. I need a way to read a blob string response as a block of memory (readonlyspan, memory, byte array, etc).  I've skimmed the code but I'm having difficulty figuring out how I might go accessing it, or even hacking together something to access it.

Thanks.

**Comment by mgravell** [comment]
Yep, you're absolutely right. All of these things are planned and intended,
just: I'm still working on other internals - specifically, I've been trying
to get a functional self-managing/monitoring connection pool going, as that
is a key transport piece.

My *plan* (of you can call it that) is to create a minimalistic but capable
redis client in a second assembly without any [assembly:InternalsVisibleTo]
magic, to prove that it all works.

But! Let's talk APIs.

Because the memory is usually directly over the underlying "pipelines"
buffers, we can't really directly expose those without jamming the pipe.
And the intricacies of multiple storage mechanisms makes it complex to
expose *anyway*.

I should definitely look into what we *can* do there to enable zero-copy
scenarios, i.e. so you can work with the data without an extra copy step.
Maybe a few TryGetBuffer style methods. However, again the emphasis here is
that most consumers won't use the RESP layer directly, and since it is
focused on alloc scenarios, it may be a little awkward to use. There's a
gnarly optimization implementation detail that means we can't always expose
things in the same way in all cases. Essentially, for most small values, we
dump the value directly into the struct rather than holding the IO buffers.


For copy-based scenarios: would it suffice to expose (all of):

- CopyTo(Span<byte>)
- CopyTo(ReadOnlySequence<byte>)
- CopyTo(Span<char>)

(and the caller provides the buffer)

Or maybe it would be better for the library to worry about where the
buffers are?

- Lifetime<ReadOnlyMemory<byte>> GetBufferCopy()

**Comment by mgravell** [comment]
I'll also see if I can add a custom value-type iterator API as a mechanism
for getting sub-items. The main question remaining being whether to also
implement IEnumerable-T, vs just relying on the language duck-typing
support (🦆 quack)

**Comment by prat0088** [comment]
@mgravell I think the copy functions would be reasonable if I could provide the buffers. If the class is reusing a buffer, which seems like an excellent optimization for 99% of the cases, then we need to do a alloc and copy at some point anyway. I don’t think my situation is so tight I can’t manage that. There will be buffer reuse internally in Respite/net code, and there will be buffer reuse in my app code, so the net effect is zero alloc, which isn’t zero copy, but it’s still very good.

In summary: copy methods where I provide the buffer will work for me. I suppose I would also need to get the length information if it is available so I could rent the destination buffer.

**Comment by mgravell** [comment]
K, I'll try and add those this weekend

**Comment by mgravell** [comment]
There now exists:

- `long RespValue.GetByteCount()`
- `int RespValue.CopyTo(Span<byte>)`
- `Lifetime<Memory<T>> Lifetime.RentMemory<T>(int)`

This should meet your needs? We may also want to consider

- `int CopyTo(ReadOnlySequence<byte>)`

but... that's for another day

**Comment by prat0088** [comment]
Tested out GetByteCount and CopyTo just now and they work as advertised.  Thank you!
